### PR TITLE
Editor: Support VideoPress post-editor preview

### PR DESCRIPTION
**Pull request by mjuhasz**
Closes: #1730

There is currently no VideoPress preview offered in the post editor on WordPress.com. When adding a VideoPress video to a post, only the shortcode text is shown in the editor post content. Only after saving and previewing the post can a user see how the post will look like.

This PR enables previewing of VideoPress videos in the wpcom-view TinyMCE plugin.

### CODE REVIEW

**Comment by nb** [comment]
@mjuhasz, this works pretty well, thanks.

Few questions:

* Few times when I pasted `[wpvideo kUJmAcSf]` in the visual editor the result had a `&nbsp;` in the middle, which broke shortcode parsing. Have you encountered this?
* The bulk of the code looks the same as the code in the gallery/contact form views. Do you think we should do something that?
* Do you think we need a stricter interface or a base class for views? (`edit`, `match`, `serialize`, etc.)
* What do you think of the overall design of `wpcom-views`? Anything you would change? How easy was it to understand how it works?

cc @aduth

**Comment by mjuhasz** [comment]
There is a sandboxing issue with the preview mode in Chrome whereby videos do not load without the `allow-same-origin` sandbox flag which the shortcode iframe lacks. It works in Safari though.

File: client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ */
+import React from 'react';
+
+/**
+ * Internal dependencies
+ */
+import shortcodeUtils from 'lib/shortcode';
+import Shortcode from 'components/shortcode';
+
+var WpVideoView = React.createClass( {
```
Comment (aduth) [inline client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx]: When writing with ES6 syntax, you shouldn't ever need to use `var`. Prefer `const` or `let` instead.

**Comment by aduth** [comment]
>Do you think we need a stricter interface or a base class for views? (`edit`, `match`, `serialize`, etc.)

Related: The embed view handler is written a bit different, splitting the component from the related logic:

https://github.com/Automattic/wp-calypso/blob/master/client/components/tinymce/plugins/wpcom-view/views/embed/index.js

>There is a sandboxing issue with the preview mode in Chrome whereby videos do not load without the `allow-same-origin` sandbox flag which the shortcode iframe lacks. It works in Safari though.

You could also try loading the VideoPress iframe directly. We do so for the previews in the media modal:

https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/media-modal/detail/detail-preview-videopress.jsx

The script loading there isn't ideal, but if I recall correctly, is primarily to manage automatic resizing of the frame as the viewport changes.

**Comment by mjuhasz** [comment]
* @nb I cannot seem to be able to reproduce the `&nbsp;` issue you described, neither in Chrome nor in Safari. Could you please confirm which browser exhibits this behavior?
* **re: duplication**: Adding this third component in its current form feels like a missed opportunity for some abstraction. As to exactly what (e.g. a base class, or slightly different design), I would have to do some read up on React, ES6, etc. to know what our options are.
* **re: onboarding**: Being completely new to the technology stack, it took me some time to set things up and figure out the flow and what to touch. It is probably easier for trained eyes. The code in general looks clean and well organized, with occasional different flavors that were confusing to me (such as `createClass` vs. inheriting from `Object`, `statics` object vs. `static` methods).
* @aduth **re: loading the VideoPress iframe directly**: Let me code that up (a skeleton at least) and see where I end up. It sounds better than making an exception for sandboxing for the `wpvideo` shortcode while shortcodes are meant to run in an unprivileged domain.

**Comment by mjuhasz** [comment]
I removed the iframe sandboxing for the wpvideo tag to make this branch fully functional. I tested another solution in PR https://github.com/Automattic/wp-calypso/pull/2267

File: client/components/shortcode/frame.jsx [hunk 2]
```diff
@@ -122,6 +123,19 @@ export default React.createClass( {
 		this.props.onLoad( event );
 	},
 
+	getSandboxAttribute() {
+		// Tags of shortcodes that do not work if rendered in a sandboxed iframe.
+		const iframeNoSandbox = [
+			'wpvideo'
+		]
+		const shortcodeTag = shortcodeUtils.parse( this.props.shortcode ).tag;
+		const trustedHost = iframeNoSandbox.some( function ( trustedShortcodeTag ) {
```
Comment (aduth) [inline client/components/shortcode/frame.jsx]: What leads you to conclude that `wpvideo` shortcodes are more trustworthy than any other shortcode?

The original concern being addressed here is that the REST API shortcode endpoint is [handled in the Jetpack plugin](https://github.com/Automattic/jetpack/blob/master/json-endpoints/class.wpcom-json-api-render-shortcode-endpoint.php), so there's a worry about the HTML being returned having originated from an external host.

**Comment by mjuhasz** [comment]
I fully agree and lean on settling on a [different approach](https://github.com/Automattic/wp-calypso/pull/2267).

**Pull request closed without merging**